src/engine/renderer/glsl_source/reliefMapping_fp.glsl

```diff
@@ -61,12 +61,16 @@ vec3 NormalInTangentSpace(vec2 texNormal)
 	normal.xy = 2.0 * normal.xy - 1.0;
 	normal.z = sqrt(1.0 - dot(normal.xy, normal.xy));
 #endif // !USE_HEIGHTMAP_IN_NORMALMAP
-
 	// HACK: 0 normal Z channel can't be good
 	if (u_NormalScale.z != 0)
 	{
 		normal *= u_NormalScale;
 	}
+
+	// HACK: prevent artifacts that may produce black normalmap pixels
+	// Unvanquished texture known to trigger the bug when doing Z reconstruction:
+	//   textures/shared_pk02_src/rock01_n
+	normal.z = clamp(normal.z, 0.0, 1.0);
```

Comment:
Apparently clamp is not well-defined for NaNs (https://www.shadertoy.com/view/Mtscz2), so we should do it a different way. Like `normal.z = sqrt(max(0, 1.0 - dot(normal.xy, normal.xy)))`